tools/gfx/cuda/render-cuda.cpp

```diff
@@ -0,0 +1,1506 @@
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        if (subObjectIndex >= objects.getCount())
+        {
+            *object = nullptr;
+            return SLANG_OK;
+        }
+        *object = objects[subObjectIndex].Ptr();
+        return SLANG_OK;
+    }
+    virtual SlangResult setObject(ShaderOffset const& offset, ShaderObject* object)
+    {
+        auto subObjectIndex =
+            layout->m_bindingRanges[offset.bindingRangeIndex].baseIndex + offset.bindingArrayIndex;
+        SLANG_ASSERT(
+            offset.uniformOffset ==
+            layout->m_bindingRanges[offset.bindingRangeIndex].uniformOffset +
+                offset.bindingArrayIndex * sizeof(void*));
+        auto cudaObject = dynamic_cast<CUDAShaderObject*>(object);
+        if (subObjectIndex >= objects.getCount())
+            objects.setCount(subObjectIndex + 1);
+        objects[subObjectIndex] = cudaObject;
+        return setData(offset, &cudaObject->bufferResource->m_cudaMemory, sizeof(void*));
+    }
+    virtual SlangResult setResource(ShaderOffset const& offset, ResourceView* resourceView)
+    {
+        auto cudaView = dynamic_cast<CUDAResourceView*>(resourceView);
+        if (offset.bindingRangeIndex >= resources.getCount())
+            resources.setCount(offset.bindingRangeIndex + 1);
+        resources[offset.bindingRangeIndex] = cudaView;
+        if (cudaView->textureResource)
+        {
+            if (cudaView->desc.type == ResourceView::Type::UnorderedAccess)
+            {
+                auto handle = cudaView->textureResource->getBindlessHandle();
+                setData(offset, &handle, sizeof(uint64_t));
+            }
+            else
+            {
+                auto handle = cudaView->textureResource->m_cudaSurfObj;
+                setData(offset, &handle, sizeof(uint64_t));
+            }
+        }
+        else
+        {
+            auto handle = cudaView->memoryResource->getBindlessHandle();
+            setData(offset, &handle, sizeof(handle));
+            auto sizeOffset = offset;
+            sizeOffset.uniformOffset += sizeof(handle);
+            auto& desc = cudaView->memoryResource->getDesc();
+            size_t size = desc.sizeInBytes;
+            if (desc.elementSize > 1)
+                size /= desc.elementSize;
+            setData(sizeOffset, &size, sizeof(size));
+
+        }
+        return SLANG_OK;
+    }
+    virtual SlangResult setSampler(ShaderOffset const& offset, SamplerState* sampler)
+    {
+        SLANG_UNUSED(sampler);
+        SLANG_UNUSED(offset);
+        return SLANG_OK;
+    }
+    virtual SlangResult setCombinedTextureSampler(
+        ShaderOffset const& offset, ResourceView* textureView, SamplerState* sampler)
+    {
+        SLANG_UNUSED(sampler);
+        setResource(offset, textureView);
+        return SLANG_OK;
+    }
```

Comment:
This case is the tricky one. If we want CUDA to seamlessly support combined texture/sampler views in this way we need to account for the combined texture/sampler pairs when computing the `ShaderObjectLayout`.

I guess in this case it shouldn't be too bad, but we'd need to make sure that the combined case also increments `m_resourceViewCount` and uses it in computing the `baseIndex`. Right now I think the logic in `ShaderObjectLayout` is acting like the combined case is using completely separate arrays.

(One longer-term question here is whether the combined case should somehow be creating and/or caching a CUDA surface object that combines the texture and sampling state provided here. It seems incorrect to just ignore the `sampler` in this case.)